consensus/src/data_manager.rs

```diff
@@ -0,0 +1,225 @@
+                    unreachable!()
+                }
+            }
+        }
+    }
+
+    pub async fn get_data(&self, block: &Block) -> Result<Vec<SignedTransaction>, Error> {
+        if block.payload().is_none() || block.payload().unwrap().is_empty() {
+            return Ok(Vec::new());
+        }
+        if self.quorum_store_enabled.load(Ordering::Relaxed) {
+            if let Payload::InQuorumStore(proofs) = block.payload().unwrap() {
+                // let data_status = self.digest_status.entry(block.id());
+                match self.digest_status.entry(block.id()) {
+                    dashmap::mapref::entry::Entry::Occupied(mut entry) => match entry.get_mut() {
+                        DataStatus::Cached(data) => Ok(data.clone()),
+                        DataStatus::Requested(receivers) => {
+                            let mut vec_ret = Vec::new();
+                            debug!("QSE: waiting for data on {} receivers", receivers.len());
+                            for rx in receivers {
+                                match rx.await {
+                                    Err(_) => {
+                                        debug!("Oneshot channel to get a batch was dropped");
+                                    }
+                                    Ok(result) => match result {
+                                        Ok(data) => {
+                                            debug!("QSE: got data, len {}", data.len());
+                                            vec_ret.push(data);
+                                        }
+                                        Err(e) => {
+                                            debug!("QS: got error from receiver {:?}", e);
+                                            let new_receivers = self
+                                                .request_data(
+                                                    proofs.clone(),
+                                                    LogicalTime::new(block.epoch(), block.round()),
+                                                )
+                                                .await;
+                                            entry.replace_entry(DataStatus::Requested(
+                                                new_receivers,
+                                            ));
+                                            return Err(e);
+                                        }
+                                    },
+                                }
+                            }
+                            let ret: Vec<SignedTransaction> =
+                                vec_ret.into_iter().flatten().collect();
+                            entry.replace_entry(DataStatus::Cached(ret.clone()));
+
+                            Ok(ret)
+                        }
+                    },
+                    dashmap::mapref::entry::Entry::Vacant(_) => {
+                        unreachable!("digest_status entry must exist!");
```

Comment:
Why is this unreachable?

IIRC you will have an entry only if you voted for the proposal. Isn't it possible to execute even if you didn't vote?

Reply:
Good point. Lets check with @zekun000.
Zekun, can we execute a block if we previously have not called execute_and_insert_block?

Reply:
AFAIU, we add all blocks to the tree before executing them (even the ones we did not vote on).
If this is true, then the entry in the DashMap can never be empty as  (1) we always call update_payload first and (2) we garbage collection only after we commit a higher block.